polkadot/xcm/xcm-fee-payment-runtime-api/tests/fee_estimation.rs

```diff
@@ -0,0 +1,366 @@
+	// Relay token is the one we want to transfer.
+	let assets = vec![(1, who, 100)]; // id, account_id, balance.
+	new_test_ext_with_balances_and_assets(balances, assets).execute_with(|| {
+		let client = TestClient;
+		let runtime_api = client.runtime_api();
+		let extrinsic = TestXt::new(
+			RuntimeCall::XcmPallet(pallet_xcm::Call::transfer_assets {
+				dest: Box::new(VersionedLocation::V4((Parent, Parachain(1000)).into())),
+				beneficiary: Box::new(VersionedLocation::V4(
+					AccountId32 { id: [0u8; 32], network: None }.into(),
+				)),
+				assets: Box::new(VersionedAssets::V4((Parent, 100u128).into())),
+				fee_asset_item: 0,
+				weight_limit: Unlimited,
+			}),
+			Some((who, extra())),
+		);
+		let dry_run_effects =
+			runtime_api.dry_run_extrinsic(H256::zero(), extrinsic).unwrap().unwrap();
+
+		assert_eq!(
+			dry_run_effects.local_program,
+			VersionedXcm::V4(
+				Xcm::builder_unsafe()
+					.withdraw_asset((Parent, 100u128))
+					.burn_asset((Parent, 100u128))
+					.build()
+			),
+		);
+
+		// In this case, the transfer type is `DestinationReserve`, so the remote xcm just withdraws
+		// the assets.
+		let send_destination = Location::new(1, Parachain(1000));
+		let send_message = Xcm::<()>::builder_unsafe()
+			.withdraw_asset((Parent, 100u128))
+			.clear_origin()
+			.buy_execution((Parent, 100u128), Unlimited)
+			.deposit_asset(AllCounted(1), [0u8; 32])
+			.build();
+		assert_eq!(
+			dry_run_effects.forwarded_messages,
+			vec![(
+				VersionedLocation::V4(send_destination.clone()),
+				VersionedXcm::V4(send_message.clone()),
+			),],
+		);
+
+		assert_eq!(
+			dry_run_effects.emitted_events,
+			vec![
+				RuntimeEvent::AssetsPallet(pallet_assets::Event::Burned {
+					asset_id: 1,
+					owner: 1,
+					balance: 100
+				}),
+				RuntimeEvent::XcmPallet(pallet_xcm::Event::Attempted {
+					outcome: Outcome::Complete { used: Weight::from_parts(200, 20) }
+				}),
+				RuntimeEvent::Balances(pallet_balances::Event::Burned { who: 1, amount: 20 }),
+				RuntimeEvent::XcmPallet(pallet_xcm::Event::FeesPaid {
+					paying: AccountIndex64 { index: 1, network: None }.into(),
+					fees: (Here, 20u128).into()
+				}),
+				RuntimeEvent::XcmPallet(pallet_xcm::Event::Sent {
+					origin: AccountIndex64 { index: 1, network: None }.into(),
+					destination: send_destination.clone(),
+					message: send_message.clone(),
+					message_id: fake_message_hash(&send_message),
+				}),
+				RuntimeEvent::System(frame_system::Event::ExtrinsicSuccess {
+					dispatch_info: DispatchInfo {
+						weight: Weight::from_parts(124414066, 0),
+						class: DispatchClass::Normal,
+						pays_fee: Pays::Yes,
+					}
+				}),
+			]
+		);
+	});
+}
+
+#[test]
+fn dry_run_xcm() {
+	let _ = env_logger::builder().is_test(true).try_init();
+	let who = 1; // AccountId = u64.
+	let transfer_amount = 100u128;
+	// TODO: We need to build the XCM to weigh it and then build the real XCM
```

Comment:
This may be OK as is. We build an XCM program that includes the `BuyExecution` with the maximum amount we are *willing* to pay. Then, we can weigh the program and learn the actual fees. If the actual fees are less than what we already provided in the `BuyExecution`, then our program is OK.

This could even be useful if the fees frequently and strongly fluctuate on a particular chain.